src/java/org/apache/cassandra/cql3/ColumnsExpression.java

```diff
@@ -0,0 +1,574 @@
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkContainsNoDuplicates;
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkContainsOnly;
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkFalse;
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkTrue;
+import static org.apache.cassandra.cql3.statements.RequestValidations.invalidRequest;
+
+/**
+ * An expression including one or several columns.
+ */
+public final class ColumnsExpression
+{
+    /**
+     * Represent the expression kind
+     */
+    public enum Kind
+    {
+        /**
+         * Single column expression (e.g. {@code columnA})
+         */
+        SINGLE_COLUMN
+        {
+            @Override
+            void validateColumns(TableMetadata table, List<ColumnMetadata> columns)
+            {
+            }
+
+            @Override
+            AbstractType<?> type(TableMetadata table, List<ColumnMetadata> columns)
+            {
+                return columns.get(0).type;
+            }
+
+            @Override
+            String toCQLString(Stream<String> columns, String mapKey)
+            {
+                return columns.findFirst().orElseThrow();
+            }
+
+            @Override
+            public String toString()
+            {
+                return "single column";
+            }
+        },
+        /**
+         * Multi-column expression (e.g. {@code (columnA, columnB)})
+         */
+        MULTI_COLUMN
+        {
+            @Override
+            protected void validateColumns(TableMetadata table, List<ColumnMetadata> columns)
+            {
+                int previousPosition = -1;
+                for (int i = 0, m = columns.size(); i < m; i++)
+                {
+                    ColumnMetadata column = columns.get(i);
+                    checkTrue(column.isClusteringColumn(), "Multi-column relations can only be applied to clustering columns but was applied to: %s", column.name);
+                    checkFalse(columns.lastIndexOf(column) != i, "Column \"%s\" appeared twice in a relation: %s", column.name, this);
+
+                    // check that no clustering columns were skipped
+                    checkFalse(previousPosition != -1 && column.position() != previousPosition + 1,
+                               "Clustering columns must appear in the PRIMARY KEY order in multi-column relations: %s", toCQLString(columns, null));
+
+                    previousPosition = column.position();
+                }
+            }
+
+            @Override
+            AbstractType<?> type(TableMetadata table, List<ColumnMetadata> columns)
+            {
+                return new TupleType(columns.stream()
```

Comment:
Don't use streams in hot path?

Reply:
once * request/s * 100k nodes... Why run the risk, the extra load and also we can't advance future uses? a for loop is equally readable. I would change it but I'll favor authorship given where it is...